### ISSUE TITLE
extension not using GPU (but deep java library example is using GPU)

### ISSUE THREAD

**Issue by kaczmarj** [qupath/qupath-extension-wsinfer#41]
i am yet again battling with GPU support on Windows. it's a real shame because it was working a few weeks ago, and now it is not. i am not sure what changed on my machine, but i'm debugging now. this may become a long post but i figured i'd document everything in case it helps anyone else (or me in a few weeks :smile:)

first i tested that a deep java library example can run and uses the gpu. i used the code below, adapted from https://docs.djl.ai/jupyter/load_pytorch_model.html. i ran all of this using the qupath script editor, after installing the pytorch djl engine via qupath. it ran and used the gpu. i monitored gpu utilization in task manager.

<details>
<summary>groovy code for running djl example</summary>

```groovy
// from https://docs.djl.ai/jupyter/load_pytorch_model.html
import java.nio.file.*;
import java.awt.image.*;
import ai.djl.*;
import ai.djl.inference.*;
import ai.djl.modality.*;
import ai.djl.modality.cv.*;
import ai.djl.modality.cv.util.*;
import ai.djl.modality.cv.transform.*;
import ai.djl.modality.cv.translator.*;
import ai.djl.repository.zoo.*;
import ai.djl.translate.*;
import ai.djl.training.util.*;

DownloadUtils.download("https://djl-ai.s3.amazonaws.com/mlrepo/model/cv/image_classification/ai/djl/pytorch/resnet/0.0.1/traced_resnet18.pt.gz", "resnet18.pt", new ProgressBar());

DownloadUtils.download("https://djl-ai.s3.amazonaws.com/mlrepo/model/cv/image_classification/ai/djl/pytorch/synset.txt", "synset.txt", new ProgressBar());

translator = ImageClassificationTranslator.builder()
        .addTransform(new Resize(256))
        .addTransform(new CenterCrop(224, 224))
        .addTransform(new ToTensor())
        .addTransform(new Normalize(
            new float[] {0.485f, 0.456f, 0.406f},
            new float[] {0.229f, 0.224f, 0.225f}))
        .optApplySoftmax(true)
        .build();
        
criteria = Criteria.builder()
        .setTypes(Image.class, Classifications.class)
        .optModelPath(Paths.get("resnet18.pt"))
        .optOption("mapLocation", "true") // this model requires mapLocation for GPU
        .optTranslator(translator)
        .optProgress(new ProgressBar()).build();

model = criteria.loadModel();

var img = ImageFactory.getInstance().fromUrl("https://raw.githubusercontent.com/pytorch/hub/master/images/dog.jpg");
predictor = model.newPredictor();

// run inference many times and watch gpu utilization.
for (x in (1..1000)) {
    classifications = predictor.predict(img);
}
```
</details>

after i found out that this worked, i tried updating the wsinfer qupath extension to add some logging and figure out what device is being used (after line https://github.com/qupath/qupath-extension-wsinfer/blob/693a576678a1ede991948d5512e9d6f3703b2c2b/src/main/java/qupath/ext/wsinfer/WSInfer.java#L164). interestingly, CPU was being used...

i will update as i debug further

**Comment by petebankhead** [comment]
Which versions of QuPath, the extension, and CUDA?

**Comment by kaczmarj** [comment]
QuPath v0.4.4. qupath wsinfer extension 0.2.0, and CUDA 11.7.

sorry i should have specified earlier.

i am making some local changes to qupath wsinfer extension to debug. will update.

**Comment by petebankhead** [comment]
Ok, QuPath v0.5.0 will be along soon and this will change things... it updates Deep Java Library to a version that uses PyTorch 2.0.1 by default. So if you don't need this to work now, it may be better to wait a week or so until after the update.

The main issue I've seen is that the version of CUDA being used is different from what PyTorch requires.

Launching QuPath from within a conda environment *might* help, and it's possible to use environment variables to change things.

Although the simplest explanation would be that you haven't expanded the additional options and selected the GPU there :)
Docs at https://qupath.readthedocs.io/en/0.4/docs/deep/wsinfer.html#additional-options

**Comment by kaczmarj** [comment]
i was confused because in the additional options, only gpu is listed. it is selected but the gpu still wasn't being used.

original qupath wsinfer extension only lists gpu (but does not use it):

![image](https://github.com/qupath/qupath-extension-wsinfer/assets/17690870/d5524c3d-2a90-4bb5-9a3b-a7b36fd3d2b7)

i modified the code to include cpu too, and now i can see cpu and gpu. cpu is set by default, and then when i select gpu, gpu is used!!! _**MASSIVE**_ speedup!

![image](https://github.com/qupath/qupath-extension-wsinfer/assets/17690870/c4cb036d-f60b-451a-8d51-e6d1740ba3e1)


indeed the following code in QuPath v0.4.4 shows that djl only lists the gpu device, as opposed to cpu. i am wondering if there is some weird behavior where having only one option does not update state for which device to use. in any case, i think making sure "cpu" is listed might fix this.

i am working on a pr

```groovy
import ai.djl.engine.Engine;

engine = Engine.getEngine("PyTorch");

for (device in engine.getDevices()) {
    println device
}
```

**Comment by petebankhead** [comment]
Ok, definitely sounds like a bug. Is it necessary to make a release with this fix for QuPath v0.4.x, or is it enough to include it in the v0.5.x release?

I can't easily test using a GPU with CUDA, but on on my Mac I was able to switch between CPU and MPS easily.

**Comment by kaczmarj** [comment]
>Is it necessary to make a release with this fix for QuPath v0.4.x, or is it enough to include it in the v0.5.x release?

no it's not necessary. i won't make the pr, but in the next release, perhaps this could be fixed. the fix might be as simple as making sure "cpu" is listed in the available devices 

```diff
        // CPU should always be available
-       if (availableDevices.isEmpty())
+       if (!availableDevices.contains("cpu"))
            availableDevices.add("cpu");
```

i was trying to debug this in the current version to iterate on examples for the manuscript.

**Comment by petebankhead** [comment]
Not a problem to make a release of v0.2.1 of the extension (compatible with QuPath v0.4) is QuPath v0.4.4 is the version to be used in the manuscript. Otherwise hoping for v0.5.0 availability next week.

I'm surprised that `cpu` isn't being returned as an available device - although I guess only because I didn't read the javadoc...
https://javadoc.io/doc/ai.djl/api/latest/ai/djl/engine/Engine.html
I expect your proposed fix should work.

**Comment by kaczmarj** [comment]
i decided to make a small pr. a release with the fix would be nice for qupath v0.4, because other windows users might be running into this issue as well.
